// @enact/ui/Measurable/Measurable.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.useMeasurable = exports.Measurable = exports["default"] = void 0;

var _hoc = _interopRequireDefault(require("@enact/core/hoc"));

var _react = _interopRequireWildcard(require("react"));

function _getRequireWildcardCache() { if (typeof WeakMap !== "function") return null; var cache = new WeakMap(); _getRequireWildcardCache = function _getRequireWildcardCache() { return cache; }; return cache; }

function _interopRequireWildcard(obj) { if (obj && obj.__esModule) { return obj; } if (obj === null || typeof obj !== "object" && typeof obj !== "function") { return { "default": obj }; } var cache = _getRequireWildcardCache(); if (cache && cache.has(obj)) { return cache.get(obj); } var newObj = {}; var hasPropertyDescriptor = Object.defineProperty && Object.getOwnPropertyDescriptor; for (var key in obj) { if (Object.prototype.hasOwnProperty.call(obj, key)) { var desc = hasPropertyDescriptor ? Object.getOwnPropertyDescriptor(obj, key) : null; if (desc && (desc.get || desc.set)) { Object.defineProperty(newObj, key, desc); } else { newObj[key] = obj[key]; } } } newObj["default"] = obj; if (cache) { cache.set(obj, newObj); } return newObj; }

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { "default": obj }; }

function _defineProperty(obj, key, value) { if (key in obj) { Object.defineProperty(obj, key, { value: value, enumerable: true, configurable: true, writable: true }); } else { obj[key] = value; } return obj; }

function _createForOfIteratorHelper(o, allowArrayLike) { var it; if (typeof Symbol === "undefined" || o[Symbol.iterator] == null) { if (Array.isArray(o) || (it = _unsupportedIterableToArray(o)) || allowArrayLike && o && typeof o.length === "number") { if (it) o = it; var i = 0; var F = function F() {}; return { s: F, n: function n() { if (i >= o.length) return { done: true }; return { done: false, value: o[i++] }; }, e: function e(_e2) { throw _e2; }, f: F }; } throw new TypeError("Invalid attempt to iterate non-iterable instance.\nIn order to be iterable, non-array objects must have a [Symbol.iterator]() method."); } var normalCompletion = true, didErr = false, err; return { s: function s() { it = o[Symbol.iterator](); }, n: function n() { var step = it.next(); normalCompletion = step.done; return step; }, e: function e(_e3) { didErr = true; err = _e3; }, f: function f() { try { if (!normalCompletion && it["return"] != null) it["return"](); } finally { if (didErr) throw err; } } }; }

function _slicedToArray(arr, i) { return _arrayWithHoles(arr) || _iterableToArrayLimit(arr, i) || _unsupportedIterableToArray(arr, i) || _nonIterableRest(); }

function _nonIterableRest() { throw new TypeError("Invalid attempt to destructure non-iterable instance.\nIn order to be iterable, non-array objects must have a [Symbol.iterator]() method."); }

function _unsupportedIterableToArray(o, minLen) { if (!o) return; if (typeof o === "string") return _arrayLikeToArray(o, minLen); var n = Object.prototype.toString.call(o).slice(8, -1); if (n === "Object" && o.constructor) n = o.constructor.name; if (n === "Map" || n === "Set") return Array.from(o); if (n === "Arguments" || /^(?:Ui|I)nt(?:8|16|32)(?:Clamped)?Array$/.test(n)) return _arrayLikeToArray(o, minLen); }

function _arrayLikeToArray(arr, len) { if (len == null || len > arr.length) len = arr.length; for (var i = 0, arr2 = new Array(len); i < len; i++) { arr2[i] = arr[i]; } return arr2; }

function _iterableToArrayLimit(arr, i) { if (typeof Symbol === "undefined" || !(Symbol.iterator in Object(arr))) return; var _arr = []; var _n = true; var _d = false; var _e = undefined; try { for (var _i = arr[Symbol.iterator](), _s; !(_n = (_s = _i.next()).done); _n = true) { _arr.push(_s.value); if (i && _arr.length === i) break; } } catch (err) { _d = true; _e = err; } finally { try { if (!_n && _i["return"] != null) _i["return"](); } finally { if (_d) throw _e; } } return _arr; }

function _arrayWithHoles(arr) { if (Array.isArray(arr)) return arr; }

/**
 * Default config for {@link ui/Measurable.Measurable}.
 *
 * @memberof ui/Measurable.Measurable
 * @hocconfig
 */
var defaultConfig = {
  /**
   * Configures the prop name to pass the measurement object.
   *
   * @type {String}
   * @default 'measurement'
   * @memberof ui/Measurable.Measurable.defaultConfig
   */
  measurementProp: 'measurement',

  /**
   * Configures the prop name to pass a function to capture a reference to the element to measure.
   *
   * @type {String}
   * @default 'forwardRef'
   * @memberof ui/Measurable.Measurable.defaultConfig
   */
  refProp: 'forwardRef'
};

var useMeasurable = function useMeasurable() {
  var _useState = (0, _react.useState)(),
      _useState2 = _slicedToArray(_useState, 2),
      measurement = _useState2[0],
      setMeasurement = _useState2[1];

  var ref = (0, _react.useRef)(null);
  (0, _react.useLayoutEffect)(function () {
    var refToMeasure = ref.current;

    if (!refToMeasure) {
      return;
    } // Support for ResizeObserver


    if (typeof ResizeObserver === 'function') {
      var resizeObserver = new ResizeObserver(function (entries) {
        entries.forEach(function (entry) {
          if (entry.target === refToMeasure) {
            // we want to measure including the padding, hence refers to `target` instead of `contentRect`
            setMeasurement(entry.target.getBoundingClientRect());
          }
        });
      });
      resizeObserver.observe(refToMeasure);
      return function () {
        resizeObserver.disconnect(refToMeasure);
        resizeObserver = null;
      }; // Fallback support for MutationObserver
    } else if (typeof MutationObserver === 'function') {
      var mutationObserver = new MutationObserver(function (mutationsList) {
        var _iterator = _createForOfIteratorHelper(mutationsList),
            _step;

        try {
          for (_iterator.s(); !(_step = _iterator.n()).done;) {
            var mutation = _step.value;

            if (mutation.type === 'childList') {
              setMeasurement(mutation.target.getBoundingClientRect());
            } else if (mutation.type === 'attributes' && (mutation.attributeName === 'style' || mutation.attributeName === 'class')) {
              setMeasurement(refToMeasure.getBoundingClientRect());
            }
          }
        } catch (err) {
          _iterator.e(err);
        } finally {
          _iterator.f();
        }
      });
      mutationObserver.observe(refToMeasure, {
        attributes: true,
        childList: true,
        subtree: true
      });
      return function () {
        mutationObserver.disconnect();
        mutationObserver = null;
      };
    }
  }, [] // disconnect on unmount
  );
  return {
    ref: ref,
    measurement: measurement
  };
};
/**
 * A higher-order component that adds the ability to measure a referenced node and get that value
 * back as a prop.
 *
 * Applying `Measurable` to a component will pass two additional props: function to fetch the `ref`
 * and the measurement object (currently this executes `getBoundingClientRect` returning its full
 * complement of properties). Both prop names are configurable through the HOC config object.
 *
 * Changes to the referenced node will result in new measurements being returned as props. That
 * being said, it probably is not wise to use this during animation as it may cause unwanted
 * performance issues.
 *
 * @see https://reactjs.org/docs/hooks-faq.html#how-can-i-measure-a-dom-node
 * @class Measurable
 * @memberof ui/Measurable
 * @hoc
 * @public
 */


exports.useMeasurable = useMeasurable;
var Measurable = (0, _hoc["default"])(defaultConfig, function (configHoc, Wrapped) {
  return function MeasurableHoc(props) {
    var _measurementProps;

    // Take the config from Measurable and insert it into a fresh instance of defineMeasurable
    var _useMeasurable = useMeasurable(),
        ref = _useMeasurable.ref,
        measurement = _useMeasurable.measurement;

    var measurementProps = (_measurementProps = {}, _defineProperty(_measurementProps, configHoc.refProp, ref), _defineProperty(_measurementProps, configHoc.measurementProp, measurement), _measurementProps);
    return /*#__PURE__*/_react["default"].createElement(Wrapped, Object.assign({}, props, measurementProps));
  };
});
exports.Measurable = Measurable;
var _default = Measurable;
exports["default"] = _default;
