// @enact/ui/AnnounceDecorator/AnnounceDecorator.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
Object.defineProperty(exports, "Announce", {
  enumerable: true,
  get: function get() {
    return _Announce["default"];
  }
});
Object.defineProperty(exports, "useAnnounce", {
  enumerable: true,
  get: function get() {
    return _useAnnounce["default"];
  }
});
exports.AnnounceDecorator = exports["default"] = void 0;

var _hoc = _interopRequireDefault(require("@enact/core/hoc"));

var _propTypes = _interopRequireDefault(require("prop-types"));

var _react = _interopRequireDefault(require("react"));

var _Announce = _interopRequireDefault(require("./Announce"));

var _useAnnounce = _interopRequireDefault(require("./useAnnounce"));

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { "default": obj }; }

function _objectWithoutProperties(source, excluded) { if (source == null) return {}; var target = _objectWithoutPropertiesLoose(source, excluded); var key, i; if (Object.getOwnPropertySymbols) { var sourceSymbolKeys = Object.getOwnPropertySymbols(source); for (i = 0; i < sourceSymbolKeys.length; i++) { key = sourceSymbolKeys[i]; if (excluded.indexOf(key) >= 0) continue; if (!Object.prototype.propertyIsEnumerable.call(source, key)) continue; target[key] = source[key]; } } return target; }

function _objectWithoutPropertiesLoose(source, excluded) { if (source == null) return {}; var target = {}; var sourceKeys = Object.keys(source); var key, i; for (i = 0; i < sourceKeys.length; i++) { key = sourceKeys[i]; if (excluded.indexOf(key) >= 0) continue; target[key] = source[key]; } return target; }

/**
 * Default config for {@link ui/AnnounceDecorator.AnnounceDecorator}.
 *
 * @memberof ui/AnnounceDecorator.AnnounceDecorator
 * @hocconfig
 * @private
 */
var defaultConfig = {
  /**
   * Configures the prop name to pass the announce function.
   *
   * @type {String}
   * @default 'announce'
   * @memberof ui/AnnounceDecorator.AnnounceDecorator.defaultConfig
   */
  prop: 'announce'
};
/**
 * A higher-order component that provides a function to its wrapped component which can be
 * called to alert the user for accessibility notifications.
 *
 * By default, the function is passed in the `announce` prop but may be customized by specifying the
 * `prop` config member. In addition to the configured prop, this HoC also adds an additional child
 * component to the passed `children` prop.
 *
 * Example:
 * ```
 * import AnnounceDecorator from '@enact/ui/AnnounceDecorator';
 * import React from 'react';
 *
 * const Example = AnnounceDecorator(class extends React.Component {
 *   static propTypes = {
 *     // passed by AnnounceDecorator
 *     announce: PropTypes.func
 *   }
 *
 *   notify = () => {
 *     const {announce} = this.props;
 *     announce('this text will be alerted to user by TTS');
 *   }
 *
 *   render () {
 *     <div>
 *       <button onClick={this.notify}>Notify on Click</button>
 *     </div>
 *   }
 * });
 * ```
 *
 * @class AnnounceDecorator
 * @memberof ui/AnnounceDecorator
 * @hoc
 * @public
 */

var AnnounceDecorator = (0, _hoc["default"])(defaultConfig, function (_ref, Wrapped) {
  var prop = _ref.prop;

  // eslint-disable-next-line no-shadow
  function AnnounceDecorator(_ref2) {
    var children = _ref2.children,
        rest = _objectWithoutProperties(_ref2, ["children"]);

    var announce = (0, _useAnnounce["default"])();
    if (prop) rest[prop] = announce.announce;
    return /*#__PURE__*/_react["default"].createElement(Wrapped, rest, children, announce.children);
  }

  AnnounceDecorator.propTypes =
  /** @lends ui/AnnounceDecorator.AnnounceDecorator.prototype */
  {
    /**
     * The wrapped component's children.
     *
     * An instance of {@link ui/AnnounceDecorator.Announce} will be appended to `children`.
     *
     * @type {Node}
     * @public
     */
    children: _propTypes["default"].node
  };
  return AnnounceDecorator;
});
exports.AnnounceDecorator = AnnounceDecorator;
var _default = AnnounceDecorator;
exports["default"] = _default;
