// @enact/ui/Scroller/Scroller.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
Object.defineProperty(exports, "ScrollerBasic", {
  enumerable: true,
  get: function get() {
    return _ScrollerBasic["default"];
  }
});
Object.defineProperty(exports, "ScrollerBase", {
  enumerable: true,
  get: function get() {
    return _UiScrollerBase["default"];
  }
});
exports.Scroller = exports["default"] = void 0;

var _propTypes = _interopRequireDefault(require("prop-types"));

var _react = _interopRequireDefault(require("react"));

var _Resizable = require("../Resizable");

var _useScroll2 = _interopRequireDefault(require("../useScroll"));

var _Scrollbar = _interopRequireDefault(require("../useScroll/Scrollbar"));

var _ScrollerBasic = _interopRequireDefault(require("./ScrollerBasic"));

var _UiScrollerBase = _interopRequireDefault(require("./UiScrollerBase"));

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { "default": obj }; }

/**
 * Unstyled scroller components and behaviors to be customized by a theme or application.
 *
 * @module ui/Scroller
 * @exports Scroller
 * @exports ScrollerBase
 * @exports ScrollerBasic
 */
var nop = function nop() {};
/**
 * An unstyled scroller.
 *
 * Example:
 * ```
 * <Scroller>Scroll me.</Scroller>
 * ```
 *
 * @class Scroller
 * @memberof ui/Scroller
 * @extends ui/Scroller.ScrollerBasic
 * @ui
 * @public
 */


var Scroller = function Scroller(props) {
  // Hooks
  var _useScroll = (0, _useScroll2["default"])(props),
      scrollContentHandle = _useScroll.scrollContentHandle,
      ScrollContentWrapper = _useScroll.scrollContentWrapper,
      isHorizontalScrollbarVisible = _useScroll.isHorizontalScrollbarVisible,
      isVerticalScrollbarVisible = _useScroll.isVerticalScrollbarVisible,
      resizeContextProps = _useScroll.resizeContextProps,
      scrollContainerProps = _useScroll.scrollContainerProps,
      scrollInnerContainerProps = _useScroll.scrollInnerContainerProps,
      scrollContentWrapperProps = _useScroll.scrollContentWrapperProps,
      scrollContentProps = _useScroll.scrollContentProps,
      verticalScrollbarProps = _useScroll.verticalScrollbarProps,
      horizontalScrollbarProps = _useScroll.horizontalScrollbarProps; // Return


  return /*#__PURE__*/_react["default"].createElement(_Resizable.ResizeContext.Provider, resizeContextProps, /*#__PURE__*/_react["default"].createElement("div", scrollContainerProps, /*#__PURE__*/_react["default"].createElement("div", scrollInnerContainerProps, /*#__PURE__*/_react["default"].createElement(ScrollContentWrapper, scrollContentWrapperProps, /*#__PURE__*/_react["default"].createElement(_ScrollerBasic["default"], Object.assign({}, scrollContentProps, {
    ref: scrollContentHandle
  }))), isVerticalScrollbarVisible ? /*#__PURE__*/_react["default"].createElement(_Scrollbar["default"], verticalScrollbarProps) : null), isHorizontalScrollbarVisible ? /*#__PURE__*/_react["default"].createElement(_Scrollbar["default"], horizontalScrollbarProps) : null));
};

exports.Scroller = Scroller;
Scroller.displayName = 'ui:Scroller';
Scroller.propTypes =
/** @lends ui/Scroller.Scroller.prototype */
{
  /**
   * A callback function that receives a reference to the `scrollTo` feature.
   *
   * Once received, the `scrollTo` method can be called as an imperative interface.
   *
   * - {position: {x, y}} - Pixel value for x and/or y position
   * - {align} - Where the scroll area should be aligned. Values are:
   *   `'left'`, `'right'`, `'top'`, `'bottom'`,
   *   `'topleft'`, `'topright'`, `'bottomleft'`, and `'bottomright'`.
   * - {node} - Node to scroll into view
   * - {animate} - When `true`, scroll occurs with animation. When `false`, no
   *   animation occurs.
   * - {focus} - When `true`, attempts to focus item after scroll. Only valid when scrolling
   *   by `node`.
   * > Note: Only specify one of: `position`, `align`, `node`
   *
   * Example:
   * ```
   *	// If you set cbScrollTo prop like below;
   *	cbScrollTo: (fn) => {this.scrollTo = fn;}
   *	// You can simply call like below;
   *	this.scrollTo({align: 'top'}); // scroll to the top
   * ```
   *
   * @type {Function}
   * @public
   */
  cbScrollTo: _propTypes["default"].func,

  /**
   * Direction of the scroller.
   *
   * Valid values are:
   * * `'both'`,
   * * `'horizontal'`, and
   * * `'vertical'`.
   *
   * @type {String}
   * @default 'both'
   * @public
   */
  direction: _propTypes["default"].string,

  /**
   * Specifies how to show horizontal scrollbar.
   *
   * Valid values are:
   * * `'auto'`,
   * * `'visible'`, and
   * * `'hidden'`.
   *
   * @type {String}
   * @default 'auto'
   * @public
   */
  horizontalScrollbar: _propTypes["default"].oneOf(['auto', 'visible', 'hidden']),

  /**
   * Prevents scroll by dragging or flicking on the scroller.
   *
   * @type {Boolean}
   * @default false
   * @private
   */
  noScrollByDrag: _propTypes["default"].bool,

  /**
   * Prevents scroll by wheeling on the scroller.
   *
   * @type {Boolean}
   * @default false
   * @public
   */
  noScrollByWheel: _propTypes["default"].bool,

  /**
   * Called when scrolling.
   *
   * Passes `scrollLeft`, `scrollTop`.
   * It is not recommended to set this prop since it can cause performance degradation.
   * Use `onScrollStart` or `onScrollStop` instead.
   *
   * @type {Function}
   * @param {Object} event
   * @param {Number} event.scrollLeft Scroll left value.
   * @param {Number} event.scrollTop Scroll top value.
   * @public
   */
  onScroll: _propTypes["default"].func,

  /**
   * Called when scroll starts.
   *
   * Passes `scrollLeft` and `scrollTop`.
   *
   * Example:
   * ```
   * onScrollStart = ({scrollLeft, scrollTop}) => {
   *     // do something with scrollLeft and scrollTop
   * }
   *
   * render = () => (
   *     <Scroller
   *         ...
   *         onScrollStart={this.onScrollStart}
   *         ...
   *     />
   * )
   * ```
   *
   * @type {Function}
   * @param {Object} event
   * @param {Number} event.scrollLeft Scroll left value.
   * @param {Number} event.scrollTop Scroll top value.
   * @public
   */
  onScrollStart: _propTypes["default"].func,

  /**
   * Called when scroll stops.
   *
   * Passes `scrollLeft` and `scrollTop`.
   *
   * Example:
   * ```
   * onScrollStop = ({scrollLeft, scrollTop}) => {
   *     // do something with scrollLeft and scrollTop
   * }
   *
   * render = () => (
   *     <Scroller
   *         ...
   *         onScrollStop={this.onScrollStop}
   *         ...
   *     />
   * )
   * ```
   *
   * @type {Function}
   * @param {Object} event
   * @param {Number} event.scrollLeft Scroll left value.
   * @param {Number} event.scrollTop Scroll top value.
   * @public
   */
  onScrollStop: _propTypes["default"].func,

  /**
   * Specifies overscroll effects shows on which type of inputs.
   *
   * @type {Object}
   * @default {drag: false, pageKey: false, wheel: false}
   * @private
   */
  overscrollEffectOn: _propTypes["default"].shape({
    drag: _propTypes["default"].bool,
    pageKey: _propTypes["default"].bool,
    wheel: _propTypes["default"].bool
  }),

  /**
   * Specifies how to scroll.
   *
   * Valid values are:
   * * `'translate'`,
   * * `'native'`.
   *
   * @type {String}
   * @default 'translate'
   * @public
   */
  scrollMode: _propTypes["default"].string,

  /**
   * Specifies how to show vertical scrollbar.
   *
   * Valid values are:
   * * `'auto'`,
   * * `'visible'`, and
   * * `'hidden'`.
   *
   * @type {String}
   * @default 'auto'
   * @public
   */
  verticalScrollbar: _propTypes["default"].oneOf(['auto', 'visible', 'hidden'])
};
Scroller.defaultProps = {
  cbScrollTo: nop,
  direction: 'both',
  horizontalScrollbar: 'auto',
  noScrollByDrag: false,
  noScrollByWheel: false,
  onScroll: nop,
  onScrollStart: nop,
  onScrollStop: nop,
  overscrollEffectOn: {
    drag: false,
    pageKey: false,
    wheel: false
  },
  scrollMode: 'translate',
  verticalScrollbar: 'auto'
};
var _default = Scroller;
exports["default"] = _default;
